lib/src/main/java/co/nimblehq/alpine/lib/mrz/MrzManager.kt

```diff
@@ -0,0 +1,145 @@
+                this.onFailure(FileNotFoundMrzProcessorException)
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            this.onFailure(DefaultMrzProcessorException)
+        }
+    }
+
+    private fun onSuccess(results: Text) {
+        scannedTextBuffer = ""
+        val textBlocks = results.textBlocks
+        if (textBlocks.isNotEmpty()) {
+            textBlocks.flatMap { textBlock ->
+                textBlock.lines.flatMap { it.elements }
+            }.forEach {
+                filterScannedText(it)
+            }
+            if (!isMrzDetected) {
+                Log.i(TAG, "MRZ not found")
+                resultListener.onError(MrzInfoNotFoundMrzProcessorException)
+            }
+        } else {
+            Log.i(TAG, "Text not found")
+            resultListener.onError(TextNotFoundMrzProcessorException)
+        }
+    }
+
+    @Suppress("MagicNumber")
+    private fun filterScannedText(element: Text.Element) {
+        scannedTextBuffer += element.text
+        val patternPassportTD3Line1 = Pattern.compile(PASSPORT_TD_3_LINE_1_REGEX)
+        val matcherPassportTD3Line1 = patternPassportTD3Line1.matcher(scannedTextBuffer)
+        val patternPassportTD3Line2 = Pattern.compile(PASSPORT_TD_3_LINE_2_REGEX)
+        val matcherPassportTD3Line2 = patternPassportTD3Line2.matcher(scannedTextBuffer)
+        if (matcherPassportTD3Line1.find() && matcherPassportTD3Line2.find()) {
+            val line2 = matcherPassportTD3Line2.group(0)
+            var documentNumber = line2?.substring(0, 9)
+            documentNumber = documentNumber?.replace("O", "0")
+            val dateOfBirth = line2?.substring(13, 19)
+            val dateOfExpiry = line2?.substring(21, 27)
+            Log.i(
+                TAG,
+                "Scanned Text Buffer Passport ->>>> Doc Number: $documentNumber " +
+                    "DateOfBirth: $dateOfBirth ExpiryDate: $dateOfExpiry"
+            )
+            val mrzInfo = MrzInfo.createFrom(documentNumber, dateOfBirth, dateOfExpiry)
+            finishScanning(mrzInfo)
+        }
+    }
+
+    private fun onFailure(e: MrzProcessorException) {
+        Log.i(TAG, "Text detection failed.$e")
+        resultListener.onError(e)
+    }
+
+    private fun finishScanning(mrzInfo: MrzInfo?) {
+        if (mrzInfo != null) {
+            with(mrzInfo) {
+                val isValidMrzInfo = documentNumber.length >= DOCUMENT_NUMBER_MINIMUM_LENGTH
+                    && dateOfBirth.length == DATE_OF_BIRTH_MINIMUM_LENGTH
+                    && dateOfExpiry.length == DATE_OF_EXPIRY_MINIMUM_LENGTH
+                if (isValidMrzInfo) {
+                    isMrzDetected = true
+                    resultListener.onSuccess(this)
+                } else {
+                    resultListener.onError(InvalidMrzInfoMrzProcessorException)
+                }
+            }
+        } else {
+            resultListener.onError(InvalidMrzInfoMrzProcessorException)
+        }
+    }
```

Comment:
How about using `run` or `apply` to reduce a heirarchy?
```suggestion
        mrzInfo?.run {
            val isValidMrzInfo = documentNumber.length >= DOCUMENT_NUMBER_MINIMUM_LENGTH
                && dateOfBirth.length == DATE_OF_BIRTH_MINIMUM_LENGTH
                && dateOfExpiry.length == DATE_OF_EXPIRY_MINIMUM_LENGTH
            if (isValidMrzInfo) {
                isMrzDetected = true
                resultListener.onSuccess(this)
            } else {
                resultListener.onError(InvalidMrzInfoMrzProcessorException)
            }
        } ?: run {
            resultListener.onError(InvalidMrzInfoMrzProcessorException)
        }
```